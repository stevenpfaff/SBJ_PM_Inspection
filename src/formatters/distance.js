const distance = (value, unit = 'mile', display = 'short') => {
  if (typeof value !== 'number' || Number.isNaN(value))
    throw new Error(
      `Invalid distance value: received '${value}' - expected number`
    )

  if (unit !== 'mile' && unit !== 'kilometer')
    throw new Error(
      `Invalid distance unit: received '${unit}' - expected 'mile' or 'kilometer'`
    )

  if (display !== 'short' && display !== 'long' && display !== 'narrow')
    throw new Error(
      `Invalid distance unit display: received '${display}' - expected 'short', 'long', or 'narrow'`
    )

  const format = Intl.NumberFormat('en-US', {
    style: 'unit',
    unit,
    unitDisplay: display
  })

  return format.format(value)
}

export default distance
