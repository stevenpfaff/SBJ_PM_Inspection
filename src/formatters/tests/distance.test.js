import distance from '../distance'

describe('distance', () => {
  it('should throw an error if typeof value is not number', () => {
    let actual
    let error

    try {
      actual = distance()
    } catch (e) {
      error = e
    }

    expect(actual).toBeUndefined()
    expect(error).toMatchSnapshot()
  })

  it('should throw an error if value is not a number', () => {
    let actual
    let error

    try {
      actual = distance(NaN)
    } catch (e) {
      error = e
    }

    expect(actual).toBeUndefined()
    expect(error).toMatchSnapshot()
  })

  it('should throw an error if unit is not "mile" or "kilometer"', () => {
    let actual
    let error

    try {
      actual = distance(1234, 'test')
    } catch (e) {
      error = e
    }

    expect(actual).toBeUndefined()
    expect(error).toMatchSnapshot()
  })

  it('should throw an error if display is not "short", "long", or "narrow"', () => {
    let actual
    let error

    try {
      actual = distance(1234, 'mile', 'test')
    } catch (e) {
      error = e
    }

    expect(actual).toBeUndefined()
    expect(error).toMatchSnapshot()
  })

  it('should return default formatted distance if only value is provided', () => {
    const expected = '1,234 mi'

    const actual = distance(1234)

    expect(actual).toBe(expected)
  })

  it('should return formatted distance if unit is "kilometer"', () => {
    const expected = '1,234 km'

    const actual = distance(1234, 'kilometer')

    expect(actual).toBe(expected)
  })

  it('should return formatted distance if display is "long"', () => {
    const expected = '1,234 miles'

    const actual = distance(1234, undefined, 'long')

    expect(actual).toBe(expected)
  })

  it('should return formatted distance if display is "narrow"', () => {
    const expected = '1,234mi'

    const actual = distance(1234, undefined, 'narrow')

    expect(actual).toBe(expected)
  })
})
