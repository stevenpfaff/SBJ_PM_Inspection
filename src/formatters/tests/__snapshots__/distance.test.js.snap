// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`distance should throw an error if display is not "short", "long", or "narrow" 1`] = `[Error: Invalid distance unit display: received 'test' - expected 'short', 'long', or 'narrow']`;

exports[`distance should throw an error if typeof value is not number 1`] = `[Error: Invalid distance value: received 'undefined' - expected number]`;

exports[`distance should throw an error if unit is not "mile" or "kilometer" 1`] = `[Error: Invalid distance unit: received 'test' - expected 'mile' or 'kilometer']`;

exports[`distance should throw an error if value is not a number 1`] = `[Error: Invalid distance value: received 'NaN' - expected number]`;
